### PR TITLE
[MC] Allocate MCFragment with a bump allocator

### PR DESCRIPTION
**Pull request by MaskRay**
#95197 and 75006466296ed4b0f845cbbec4bf77c21de43b40 eliminated all raw
`new MCXXXFragment`. We can now place fragments in a bump allocator.
In addition, remove the dead `Kind == FragmentType(~0)` condition.

~CodeViewContext may call `StrTabFragment->destroy()` and need to be
reset before `FragmentAllocator.Reset()`.
Tested by llvm/test/MC/COFF/cv-compiler-info.ll using asan.

### CODE REVIEW

File: llvm/include/llvm/MC/MCContext.h [hunk 1]
```diff
@@ -136,6 +136,9 @@ class MCContext {
   /// objects.
   BumpPtrAllocator Allocator;
 
+  /// For MCFragment instances.
+  BumpPtrAllocator FragmentAllocator;
```
Comment (aengelke) [inline llvm/include/llvm/MC/MCContext.h:140]: Any particular reasons to not reuse `Allocator`?

File: llvm/lib/MC/MCFragment.cpp [hunk 2]
```diff
@@ -203,59 +203,57 @@ MCFragment::MCFragment(FragmentType Kind, bool HasInstructions)
 
 void MCFragment::destroy() {
   // First check if we are the sentinel.
-  if (Kind == FragmentType(~0)) {
```
Comment (aengelke) [inline llvm/lib/MC/MCFragment.cpp:206]: `~MCFragment();`?

**Review by aengelke — APPROVED** [review]
LGTM, with two nits/questions?

File: llvm/include/llvm/MC/MCContext.h (hunk 1 above)
Comment (MaskRay) [reply llvm/include/llvm/MC/MCContext.h:140]: Make fragments closer for locality.

In addition, there might be an opportunity to reduce `MCFragment::destroy` (primarily called by `~MCSection`) overhead.
If we can obtain BumpPtrAllocator slabs, there will probably be a better way to implement `-mc-disable-free` like `-cc1 -disable-free`.

Comment (aengelke) [reply llvm/include/llvm/MC/MCContext.h:140]: Good point, ok

File: llvm/lib/MC/MCFragment.cpp (hunk 2 above)
Comment (MaskRay) [reply llvm/lib/MC/MCFragment.cpp:206]: It looks like the `Kind == FragmentType(~0)` code path is now dead. I'll just delete this condition.

**Comment by nikic** [comment]
It looks like this change has massively increased memory usage, see https://llvm-compile-time-tracker.com/compare.php?from=fc23564c44f3eff1847462253d43c08b85489148&to=8cb6e587fd40b97983e445ee3f17f4c6d7190df7&stat=max-rss&linkStats=on. ReleaseLTO-g linking max-rss goes up by up to 7%.

**Comment by MaskRay** [comment]
Interesting, so there is tiny instruction:u decrease with some max-rss increase, significant for `stage1-ReleaseLTO-g (link only)`.

This increase likely stems from LTO optimizations allocating objects on the heap that overlap significantly with newly created MCDataFragment instances.
The bump allocator eliminates this overlap and increases max-rss.
This can be mitigated if we can identify the heap objects and switch the optimization pass(es) to bump allocators.

This bump allocator change opens doors for further performance (e.g. more elegant -disable-free for fragments) and max-rss improvement, so I want to keep it despite the ReleaseLTO max-rss regression...
We can view this as using the previous saved credit from max-rss optimizations to unlock future gains...

**Comment by nikic** [comment]
> This increase likely stems from LTO optimizations allocating objects on the heap that overlap significantly with newly created MCDataFragment instances. The bump allocator eliminates this overlap and increases max-rss. This can be mitigated if we can identify the heap objects and switch the optimization pass(es) to bump allocators.

So you're saying that there is no increase in memory usage per se, and the max-rss increase is because the allocator cannot reuse already mapped memory (e.g. because it is reserved for smaller type classes, or whatever)? Have you verified that this is the case with massif?

Note that the max-rss increase is not limited to LTO only, it is also seen at `-O0`, to a lesser degree. Backend memory usage increases are just always easiest to see in the LTO configuration, as backend tends to dominate max-rss there.

> This bump allocator change opens doors for further performance (e.g. more elegant -disable-free for fragments) and max-rss improvement, so I want to keep it despite the ReleaseLTO max-rss regression...

I think this regression may be too large to keep it for speculative future gains. LTO memory usage is a pretty big problem, and the increase here is not small.

Regarding `-disable-free`, that would not work for frontends that codegen on multiple threads, right? Including rustc and lld with ThinLTO. If so, I'm not sure this is a useful optimization.

**Comment by aengelke** [comment]
I'm admittedly quite surprised about the max-rss increase -- that shouldn't happen (at least not to such an extent), because fragments are still freed when sections are freed, which is done by MCContext::reset(), which will free the allocated memory for fragments. Sections are freed before fragments (at least, if DoAutoReset is set, which it looks like it is? If it's not, that might be problematic...), so fragment destructors should get called and free associated contents/fixups. I guess we are overlooking something. Will see if I find time to look at this tomorrow.

**Pull request merged**